llvm/include/llvm/Analysis/ScalarEvolutionPatternMatch.h

```diff
@@ -192,15 +193,18 @@ struct SCEVBinaryExpr_match {
 
   bool match(const SCEV *S) const {
     auto *E = dyn_cast<SCEVTy>(S);
-    return E && E->getNumOperands() == 2 && Op0.match(E->getOperand(0)) &&
-           Op1.match(E->getOperand(1));
+    return E && E->getNumOperands() == 2 &&
+           ((Op0.match(E->getOperand(0)) && Op1.match(E->getOperand(1))) ||
+            (Commutable && Op0.match(E->getOperand(1)) &&
+             Op1.match(E->getOperand(0))));
```

Comment:
Wait a minute: why do we have Commutable? Aren't SCEV expressions canonical, and therefore, there being only one possibility for the order of the operands?

Reply:
That is true but it is dependent on what the operands are, which the caller doesn't know.  The specific issue is that constants are canonicalised to be the first operand.  This means the expected canonical forms are either `constant * vscale` or `vscale * variable`.